Aot watch

tbosch opened the pull request.

tbosch committed:
refactor(compiler): move `static_reflector` into `@angular/compiler` and rename files

- `src/runtime_compiler.ts` -> `src/jit/compiler.ts`
- `src/compiler.ts` -> `src/jit/compiler_factory.ts`
- `src/offline_compiler` -> `src/aot/compiler.ts`

Part of #12867

tbosch committed:
refactor(compiler): move `findDeclaration` into the `StaticReflector`

Previously, this was part of the `AotCompilerHost`.
The `AotCompilerHost` is now also greatly simplified.

tbosch committed:
refactor(compiler): remove `asset:` urls

These urls were just relicts from Dart.

tbosch committed:
refactor(compiler): add `createAotCompiler` factory

Also adds 2 more methods to the `AotCompilerHost`:
- `loadResource`
- `resolveFileToImport`

tbosch committed:
refactor(compiler): move symbol extraction to `AotCompiler`

tbosch committed:
fix(compiler): fix versions of `@angular/tsc-wrapped`

tbosch committed:
refactor(tsc-wrapped): collect all exported functions and classes and bump metadata version from 1 to 2

This is needed to resolve symbols without `.d.ts` files.
This bumps the version of the metadata from 1 to 2.
This adds logic into `ng_host.ts` to automatically upgrade
version 1 to version 2 metadata by adding the exported symbols
from the `.d.ts` file.

alexeagle commented on modules/@angular/compiler-cli/src/ng_host.ts:
```diff
@@ -75,7 +76,7 @@ export class NgHost implements AotCompilerHost {
    * NOTE: (*) the relative path is computed depending on `isGenDirChildOfRootDir`.
    */
-  getImportPath(containingFile: string, importedFile: string): string {
+  resolveFileToImport(importedFile: string, containingFile: string): string {
```
fileToRelativeImport?
https://nodejs.org/api/path.html#path_path_relative_from_to

I would prefer these APIs match my understanding from Node's path handling

tbosch replied:
Changed to `moduleNameToFileName` and `fileNameToModuleName`.

alexeagle commented on modules/@angular/compiler-cli/test/ng_host_spec.ts:
```diff
@@ -57,87 +56,87 @@ describe('NgHost', () => {
     it('should import application from factory', () => {
-      expect(hostNestedGenDir.getImportPath(
-                 '/tmp/project/src/my.ngfactory.ts', '/tmp/project/src/my.other.ts'))
+      expect(hostNestedGenDir.resolveFileToImport(
```
I'd like the order to match node: from, to

alexeagle commented on tools/@angular/tsc-wrapped/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@angular/tsc-wrapped",
-  "version": "0.2.2",
+  "version": "0.4.1",
```
0.4.0 still until release

tbosch replied:
Done

alexeagle commented on modules/@angular/compiler-cli/src/ng_host.ts:
```diff
@@ -150,43 +150,69 @@ export class NgHost implements AotCompilerHost {
+        }
+        const sourceText = this.context.readFile(dtsFilePath);
+        const exports = this.metadataCollector.getMetadata(
+            ts.createSourceFile(dtsFilePath, sourceText, ts.ScriptTarget.Latest, true));
```
first look for it in the program?

tbosch replied:
Done

tbosch committed:
refactor(compiler): renames

- `NgHost` to `CompilerHost`
- `AotCompilerHost.resolveFileToImport` to `AotCompilerHost.fileNameToModuleName`
- `AotCompilerHoset.resolveImportToFile` to `AotCompilerHost.moduleNameToFileName`

tbosch commented on modules/@angular/compiler/src/aot/compiler_host.ts:
```diff
@@ -0,0 +1,31 @@
+  getMetadataFor(modulePath: string): {[key: string]: any}|{[key: string]: any}[];
+
+  /**
+   * Converts a module name into a file path.
```
Added.

tbosch commented on modules/@angular/compiler-cli/src/codegen.ts:
```diff
@@ -17,9 +17,9 @@ import {readFileSync} from 'fs';
 import * as ts from 'typescript';
 
-import {PathMappedReflectorHost} from './path_mapped_reflector_host';
+import {NgHost, NgHostContext} from './ng_host';
```
Done

tbosch committed:
refactor(comiler): various cleanups

tbosch committed:
refactor(compiler): Reintroduce `ReflectorHost` and move `Extractor` into `@angular/compiler`

tbosch committed:
refactor(compiler): further minor fixes

tbosch commented:
Needs to be synced into G3 with cl/139588792

tbosch commented:
Reviewed in person with @alexeagle
